arc.m

```diff
@@ -767,39 +773,47 @@ static BOOL setObjectHasWeakRefs(id obj)
 OBJC_PUBLIC id objc_storeWeak(id *addr, id obj)
 {
 	LOCK_FOR_SCOPE(&weakRefLock);
-	WeakRef *oldRef;
-	id old;
-	loadWeakPointer(addr, &old, &oldRef);
-	// If the old and new values are the same, then we don't need to do anything.
-	if (old == obj)
+	BOOL assignNoAndUnregister = NO;
+	if(obj == nil)
 	{
-		return obj;
+		assignNoAndUnregister = YES;
 	}
 	BOOL isGlobalObject = setObjectHasWeakRefs(obj);
-	// If we old ref exists, decrement its reference count.  This may also
-	// delete the weak reference control block.
-	if (oldRef != NULL)
-	{
-		weakRefRelease(oldRef);
-	}
-	// If we're storing nil, then just write a null pointer.
-	if (nil == obj)
-	{
-		*addr = obj;
-		return nil;
-	}
 	if (isGlobalObject)
 	{
 		// If this is a global object, it's never deallocated, so secretly make
 		// this a strong reference.
 		*addr = obj;
-		return obj;
+		return *addr;
 	}
-	if (nil != obj)
+	if(!assignNoAndUnregister)
 	{
-		*addr = (id)incrementWeakRefCount(obj);
+		// obj is not null, update it's registration
+		WeakRef *oldRef;
+		id old;
+		loadWeakPointer(addr, &old, &oldRef);
+		// If the old and new values are the same, then we don't need to do anything.
+		if (old == obj && obj != nil)
+		{
+			return *addr;
+		}
+		// release old ref
+		if (oldRef != NULL) {
+			weakRefRelease(oldRef);
+		}
+		// set new ref
+		*(WeakRef**)addr = (id)incrementWeakRefCount(obj);
+		return obj;
 	}
-	return obj;
+	else {
+		*addr = nil;
+		WeakRef *oldRef;
+		id old;
+		if(loadWeakPointer(addr, &old, &oldRef)) {
+			weakRefRelease(oldRef);
+		}
+	}
+	return *addr;
```

Comment:
The new logic is harder to follow, but I'm struggling to see what's actually different here.  Both versions are releasing the old weak ref if they find one (you're doing that on both code paths) and if the new value is not nil then incrementing its weak ref count and storing it in the variable.